### PR TITLE
Precompute some concretization information

### LINKED ISSUE #244: Cache reusable data to reduce latency of getConcretizationInfo
#215 allows us to hold a dynamic `Fusion`, and concretize it as needed based on provided inputs. When we see new inputs, we must determine whether we can reuse any of the previous concretizations. To do so, we call `getConcretizationInfo` and use the resulting structure in a key for a hash table. This can be a somewhat heavy call, as it involves traversing the dynamic `Fusion` to find reshape transforms and propagation of scalars using `ExpressionEvaluator`.  Both of these operations start from scratch each time.

In order to speed up subsequent calls, we could instead pre-compute some of this information.
- Using the dynamic `Fusion`, build an `ExpressionEvaluator` that evaluates as much as possible ahead of receiving inputs.
- Find a list of reshape expressions
- Register this as Fusion-managed data (#45)
This would then allow us to copy the dynamic Fusion which would have some of the concretization work already done for us, and should speed up the computation of `getConcretizationInfo`.

### PR DESCRIPTION
**Pull request by jacobhinkle**
Fixes #244.

Introduces `DynamicTransformInitialInfo` which is information obtained before inputs are available that is used to speed up the actual concretization information. In particular, graph traversals to find dynamic ops are only done once, to build this object, after which each new set of inputs that needs to be concretized only needs to replace those ops (constant time) followed by traversals to update symbolic IterDomain types and update TV uses.

The `DynamicTransformInitialInfo` object also holds a _minimal_ set of input scalars that determine concretization. This is used for minimizing the cache ID associated with `KernelArgumentHolder` so that input scalars which do not affect concretization do not mistakenly alter the key.

We also attach an `ExpressionEvaluator` to `DynamicTransformInitialInfo`, and warm it up so that concretization is slightly faster.

Incidentally, this PR also adds a new `fusion_ir_concretized` debug dump option.

### CODE REVIEW

File: csrc/expr_evaluator.cpp [hunk 1]
```diff
@@ -232,4 +232,17 @@ void ExpressionEvaluator::propagateBoundValuesThroughExactMaps(Fusion* fusion) {
...
+      !precomputed_values_,
+      "Cannot clone ExpressionEvaluator with bound PrecomputedValues");
+  for (const auto& kv : known_values_) {
+    expr_eval.known_values_[ir_cloner.clone(kv.first)] = kv.second;
+  }
+  expr_eval.known_named_scalars_.insert(
+      known_named_scalars_.begin(), known_named_scalars_.end());
+  return expr_eval;
+}
+
```
Comment (jacobhinkle) [inline csrc/expr_evaluator.cpp:247]: These are needed in order to clone the partially-evaluated expr_eval to the new `Fusion` which will be concretized.

**Comment by naoyam** [comment]
!build

File: csrc/dynamic_transform.cpp [hunk 2]
```diff
@@ -20,43 +20,253 @@
...
+    // initial_info_ provides a set of input Vals that are used for
+    // concretization. Here we check which inputs, if any, correspond to any of
+    // those Vals. These will be the inputs that are explicitly used in the
+    // cache Id for KernelArgumentHolder.
+    info_.input_affects_concretization_.resize(fusion->inputs().size(), false);
+    auto dyn_vals = info_.getRootDynamicVals();
+    for (const auto i : c10::irange(fusion->inputs().size())) {
+      auto input = fusion->inputs().at(i);
+      info_.input_affects_concretization_[i] =
+          input->isA<TensorView>() || dyn_vals.find(input) != dyn_vals.end();
```
Comment (naoyam) [inline csrc/dynamic_transform.cpp:84]: Why any input is considered affecting concretization?

File: csrc/kernel_cache.h [hunk 3]
```diff
@@ -283,7 +283,7 @@ class TORCH_CUDA_CU_API InputsIdLookup : public NonCopyable {
   //! depend on those inputs and omitting them would lead to a collision.
   IdLookupReturn lookupId(
       const at::ArrayRef<c10::IValue>& inputs,
-      bool hash_scalars = false);
+      const std::vector<bool>* input_affects_concretization = nullptr);
```
Comment (naoyam) [inline csrc/kernel_cache.h:286]: Please update the function comment above.

Comment (naoyam) [reply csrc/kernel_cache.h:286]: Why is this a pointer to a vector? Why nullptr is allowed?

File: csrc/kernel_cache.cpp [hunk 4]
```diff
@@ -283,8 +298,8 @@ KernelArgumentHolder FusionExecutorCache::prepareInputs(
   // fusions. This may not be ideal in all cases, since it will prevent
   // short-circuiting here, resulting in avoidable rebuilds of concretization
   // info.
-  auto id_lookup_ret =
-      inputs_id_lookup_.lookupId(inputs, /*hash_scalars*/ isDynamic());
+  auto id_lookup_ret = inputs_id_lookup_.lookupId(
+      inputs, &(initialInfo().inputsAffectConcretization()));
```
Comment (naoyam) [inline csrc/kernel_cache.cpp:302]: Not quite sure if this is safe. Looks like a pointer to a temporary is passed to `lookupId`. Isn't the temporary invalid when `lookupId` is executed? Or, does C++ guarantee the lifetime of the temporary is extended like its const reference is passed?

Either way, just passing its const reference seems simpler and better.

File: csrc/dynamic_transform.cpp [hunk 5]
```diff
@@ -20,43 +20,253 @@
...
 };
 
-DynamicTransformInfoBuilder::DynamicTransformInfoBuilder(
-    Fusion* fusion,
-    ExpressionEvaluator* expr_eval)
-    : expr_eval_(expr_eval), info_(fusion) {
-  TORCH_INTERNAL_ASSERT(
-      !fusion->isA<kir::Kernel>(),
-      "Invalid container. Kernel container not allowed.\n");
+void DynamicTransformConcretizationInfo::analyzeReshapes(
```
Comment (naoyam) [inline csrc/dynamic_transform.cpp:154]: Just to confirm, is this just moved? Same for `analyzeResizes`.

**Review by naoyam — COMMENTED** [review]
Looks good overall. A couple of minor things to clarify.

**Comment by naoyam** [comment]
With this PR, what are the additional steps incurred when dynamic reshape or resize is used?

**Comment by jacobhinkle** [comment]
> With this PR, what are the additional steps incurred when dynamic reshape or resize is used?

Previously, in the constructor of `FusionExecutorCache` we called `fusion->hasDynamicTransform()` to perform a traversal to find any TVs with `Symbolic` IDs, caching the result. First, we move that to the first call to `runFusionWithInputs` since the Python frontend actually defines the `Fusion` after constructing `FusionExecutorCache`, so `fusion` is empty inside the constructor. This move would be needed with or without this PR once the frontend is updated to support the dynamic ops.

We also replace that call to `hasDynamicTransform()` with a traversal using `DynamicTransformInitialInfoBuilder` that collects the dynamic resize and reshape ops; after this, it suffices to check whether those op lists are empty to determine if the `Fusion` is static. It also initializes an `ExpressionEvaluator` and attempts to compute important scalars like the input or output extents of reshape and resize ops, so that intermediate values will be cached. `DynamicTransformInitialInfoBuilder` also collects a set of input scalars that appear in those expressions, which gives us a minimal list of inputs to create our highest level cache with. Given inputs, if that cache misses we compute the `DynamicTransformConcretizationInfo` using the `DynamicTransformInitialInfo` along with the inputs. This does not require another traversal to find dynamic ops since they were recorded in the initial info, so we can do those replacements then do a single traversal to propagate the concretized IDs in the p2c direction.

**Comment by naoyam** [comment]
Thanks. I think there are three major times to consider:

1. Initialization time
2. Cache lookup time
3. Upon cache miss

Can you rephrase what you mentioned above to clarify what's done at each step?

**Comment by jacobhinkle** [comment]
> Can you rephrase what you mentioned above to clarify what's done at each step?

Sure. There are two levels of cache implemented by `FusionExecutorCache`, the first with key `KernelArgumentHolder::getCacheId()` based on the input scalars and the second keyed on concretization info. So I added the second item below:

1. **Initialization of `FusionExecutorCache`**
This PR saves one traversal here by not calling `hasDynamicTransforms()`. The constructor of `FusionExecutorCache` is now essentially empty.

2. **Building `KernelArgumentHolder` cache ID**
*(First time only)* This PR adds a traversal to find initial information and warm up expression evaluator. A traversal similar to this `fusion->hasDynamicTransforms()` was previously performed in step 1 above.
*(Every subsequent call)* This PR doesn't significantly change speed of creating the input scalar cache key; it merely reduces the amount of information there to encourage more hits.

3. **Building concretization info cache ID (i.e. `KernelArgumentHolder` cache miss)**
This PR reduces processing required to build concretization info at this stage since no traversal is required here; we can visit dynamic ops directly since they are recorded in the "initial info".

4. **Concretization info cache miss**
In this case, we need to concretize the `Fusion`, then schedule/compile it. This PR does not change that process much at all; a traversal is required to propagate concretized dynamic ops downstream.

**Comment by naoyam** [comment]
Thanks! That's really helpful information. Can you please update this comment?

https://github.com/NVIDIA/Fuser/blob/main/csrc/kernel_cache.h#L326-L367

**Comment by jacobhinkle** [comment]
> this comment

Ah of course! Thanks for catching that. I will update this comment to reflect the current situation.

File: csrc/kernel_cache.h (hunk 3 above)
Comment (jacobhinkle) [reply csrc/kernel_cache.h:306]: Added comment to explain this argument. `nullptr` is allowed since lookupId is also used for a higher-level cache in the frontend which cannot know what inputs affect concretization, so must include them all: https://github.com/NVIDIA/Fuser/blob/main/csrc/python_frontend/fusion_cache.cpp#L203.

File: csrc/kernel_cache.h [hunk 6]
```diff
@@ -323,57 +343,71 @@ class TORCH_CUDA_CU_API InputsIdLookup : public NonCopyable {
...
+//! (in the form of a DynamicTransformConcretizationInfo object) to a vector of
+//! FusionKernelRuntimes, and check whether each one is able to run the present
+//! inputs. If not, we create a new FusionKernelRuntime for those inputs and
+//! record it in the list (as well as recording a mapping from the input ID to
+//! the new runtime).
 //!
+//! In the case of a dynamic Fusion, input scalars such as integer parameters to
+//! a model could potentially affect the structure of a concretized Fusion, so
+//! we take care to include those scalars in the input ID along with the extents
+//! of tensor arguments.
```
Comment (jacobhinkle) [inline csrc/kernel_cache.h:404]: ^ Updated note on "post-definition" caching

Comment (jacobhinkle) [reply csrc/kernel_cache.h:404]: Ah of course. I was focusing on `Val`s and forgot to replace that. Will update.

Comment (jacobhinkle) [reply csrc/kernel_cache.h:404]: Updated to place the original note back in place, and also updated my new note for correctness about additional tensor properties other than shape.

File: csrc/dynamic_transform.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:84]: It's only considered to affect concretization if it's in `dyn_vals` or it's a `TensorView`, in which case the value `input_affects_concretization_[i]` is ignored and the extents/contiguity/dtype/etc are used in the ID string.

File: csrc/kernel_cache.cpp (hunk 4 above)
Comment (jacobhinkle) [reply csrc/kernel_cache.cpp:302]: The const reference is indeed better if the argument is not optional. I used a pointer instead to handle a default value, but that's actually more trouble than it's worth. I'll change it to a const reference with an override for when the argument is missing.

File: csrc/dynamic_transform.cpp (hunk 5 above)
Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:154]: Yes, these were previously in the `DynamicTransformInfoBuilder::handle()` overrides. I just copied them to here and update to use an explicit `ExpressionEvaluator`. I don't keep `expr_eval` inside `DynamicTransformConcretizationInfo` since I use it with `std::hash` so I'd like to keep it small. Instead I copy the `ExpressionEvaluator` from the initial info in `getConcretizationInfo` and pass it to these functions.

File: csrc/dynamic_transform.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/dynamic_transform.cpp:84]: Sorry, I meant to ask why any tensor is considered affecting concretization. I think my confusion was because of the name. All tensors are used for the cache lookup, but not necessarily for fusion concretization. Wouldn't it make more sense to use a different name that would indicate these are used as a key for caching?

File: csrc/dynamic_transform.cpp [hunk 7]
```diff
@@ -20,43 +20,253 @@
...
+    const auto inputs = InputsOf::outputs(info_.fusion(), leaf_dynamic_vals_);
+    info_.root_dynamic_vals_.insert(inputs.begin(), inputs.end());
   }
 
  private:
-  ExpressionEvaluator* expr_eval_ = nullptr;
+  DynamicTransformInitialInfo info_;
 
-  DynamicTransformConcretizationInfo info_;
+  std::vector<Val*> leaf_dynamic_vals_;
```
Comment (naoyam) [inline csrc/dynamic_transform.cpp:151]: Can you add a short comment about this?

File: csrc/dynamic_transform.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:84]: Ah yes. I agree it's a weird name, and tensorviews are pretty much ignored, so I could use an `unordered_set`. However, since this is the fastest path I figured checking a vector of bools would be faster than searching for each scalar. That concern is probably secondary to clarity though, so maybe it'd be better to switch to `unordered_set<Val*> concretization_affecting_scalars` or some similar name.

Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:84]: Renamed to `record_scalar` which is more explicit for `lookupId`. I didn't switch to `unordered_set<Val*>` since `lookupId` works with a vector of `c10::IValue`s, so we'd need to zip with the fusion inputs to figure out which vals to record anyway. We could instead record their positions as ints, but I figured that is just about as complex as creating a bool vector as we do now.

File: test/test_gpu2.cpp [hunk 8]
```diff
@@ -1070,7 +1070,7 @@ TEST_F(NVFuserTest, FusionInputsIdLookup_CUDA) {
   nvfuser::InputsIdLookup inputs_id_lookup(2);
 
   // testing basic function, same encoding for identical inputs
-  auto id_0 = inputs_id_lookup.lookupId({t0, t1, 5.0});
+  auto id_0 = inputs_id_lookup.lookupId({t0, t1, 2.5});
```
Comment (jacobhinkle) [inline test/test_gpu2.cpp:1073]: @kevinstephano FYI `lookupId` now uses all scalar values to create ID by default. This is because with dynamic shapes, any scalar value could potentially affect the structure of the concretized Fusion, and we cannot know which ones do affect it until the Fusion is built. See below in this test for example of masking scalar args out with the `record_scalar` vector argument.

File: csrc/kernel_cache.cpp [hunk 9]
```diff
@@ -44,9 +44,12 @@ c10::ThreadPool* getThreadPool() {
   return &pool;
 }
 
-void encodeBuffer(size_t value, std::string& buffer) {
+// Copy bytes of value to back of buffer. This is templated in order to avoid
+// implicit cast such as int64_t -> size_t that might lose information.
+template <typename T>
+void encodeBuffer(T value, std::string& buffer) {
```
Comment (jacobhinkle) [inline csrc/kernel_cache.cpp:50]: I noticed that a signed int scalar input might now be passed to this function resulting in an implicit cast to `size_t`. In order to ensure we are always appending the input byte-for-byte I templated it.

File: csrc/dynamic_transform.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/dynamic_transform.cpp:84]: `info_.input_affects_concretization_` still sounds a little confusing me. Again, this is primarily for cache lookup, not fusion concretization. A tensor input may have `true` here, but it doesn't mean it affects concretization.

Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:84]: When passed to `lookupId`, you're right that these are just the scalars that should be recorded (or the TensorViews). The entries for each input position are `true` iff either the input is a tensor or it appears in some scalar expression that we check during concretization (i.e. something in `leaf_dynamic_vals_`), so my intention was that inside `DynamicTransformInitialInfo` we actually _are_ tracking just the inputs that affect concretization.

Comment (naoyam) [reply csrc/dynamic_transform.cpp:84]: > we actually are tracking just the inputs that affect concretization.

Is that true for tensor inputs?

Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:84]: Touche. It's ignored in `lookupId` anyway, but for consistency we can probably just set `TensorView` input entries to `false` for clarity.

Comment (naoyam) [reply csrc/dynamic_transform.cpp:84]: Well, some input tensors do affect concretization.

Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp:84]: OK, I'll check input TensorViews. If their extents affect concretization they will have a true value as well; otherwise false.

Comment (naoyam) [reply csrc/dynamic_transform.cpp:84]: IIUC, the only information we need is which scalar inputs are necessary for cache lookup. We could just have `unordered_set<int64_t> scalar_input_affect_concretization` to hold the indices of the scalar inputs in the input vector.

File: test/test_gpu2.cpp (hunk 8 above)
Comment (jacobhinkle) [reply test/test_gpu2.cpp]: Nevermind. I actually updated this to default back to the behavior assumed here, so there should be no change in behavior for user schedules.

File: csrc/dynamic_transform.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp]: Switched to `unordered_set`. As you said I think this is the clearest approach, and in many use cases the set will be empty so the check is cheap.

**Review by naoyam — APPROVED** [review]
LGTM.
